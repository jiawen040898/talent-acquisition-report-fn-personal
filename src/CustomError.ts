export enum ErrorType {
    INTERNAL_SERVER_ERROR = 'InternalServerError',
    TIMED_OUT = 'TimedOut',
    FILE_NOT_FOUND = 'FileNotFound',
}

export enum ErrorMessage {
    INTERNAL_SERVER_ERROR = 'Something went wrong, please contact support.',
    TIMED_OUT = 'Request took longer than expected.',
    FILE_NOT_FOUND = 'File not found',
}

export enum ErrorCode {
    FILE_NOT_FOUND = 'file_not_found',
}

export class CustomError extends Error {
    name: string;
    statusCode: number;

    constructor(
        message: string | SafeAny,
        name = 'Unknown Error',
        error: Error | null = null,
        statusCode = 400,
    ) {
        if (typeof message === 'object') {
            message = JSON.stringify(message, null, 2);
        }
        super(message);
        this.name = name;
        this.statusCode = statusCode;
        this.stack = error?.stack;
    }

    static get(
        errorType: ErrorType,
        error: Error | null = null,
        isBadRequest = true,
    ): CustomError {
        let errorMessage = ErrorMessage.INTERNAL_SERVER_ERROR;
        switch (errorType) {
            case ErrorType.INTERNAL_SERVER_ERROR:
                errorMessage = ErrorMessage.INTERNAL_SERVER_ERROR;
                break;
            case ErrorType.TIMED_OUT:
                errorMessage = ErrorMessage.TIMED_OUT;
                break;
        }

        return new CustomError(
            errorMessage,
            errorType,
            error,
            isBadRequest ? 400 : 500,
        );
    }
}
