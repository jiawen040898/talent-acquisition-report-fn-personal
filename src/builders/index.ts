export * from './base-query.builder';
export * from './interview-rating-query.builder';
export * from './job-application-query.builder';
