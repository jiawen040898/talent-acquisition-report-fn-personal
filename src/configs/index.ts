export * from './aws.config';
export * from './pulsifi.config';
export * from './translation.config';
export * from './unleash.config';
