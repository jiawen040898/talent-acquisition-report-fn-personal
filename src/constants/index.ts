export * from './app-notification.enum';
export * from './cache-store.enum';
export * from './channel.enum';
export * from './company-status.enum';
export * from './contact';
export * from './email-type.enum';
export * from './errors.enum';
export * from './feature-toggle-name.enum';
export * from './fit-score-recipe.enum';
export * from './icons';
export * from './interview-rating.enum';
export * from './interview-template';
export * from './job.enum';
export * from './job-application';
export * from './job-application-channel.enum';
export * from './job-application-score-type.enum';
export * from './job-application-screening-answer-criteria-status.enum';
export * from './job-application-skill-proficiency.enum';
export * from './job-distribution-score-type.enum';
export * from './namespace';
export * from './psychology-api-errors.enum';
export * from './report-download-notification-type.enum';
export * from './screening-question.enum';
export * from './sort.enum';
export * from './styles';
export * from './template';
export * from './unleash-errors.enum';
