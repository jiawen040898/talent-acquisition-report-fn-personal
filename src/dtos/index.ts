export * from './fit-score-recipe.dto';
