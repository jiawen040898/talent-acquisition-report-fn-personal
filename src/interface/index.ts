export * from './assessment-report-request';
export * from './company.interface';
export * from './download-report-webhook-event.interface';
export * from './email-request';
export * from './generator';
export * from './handlebar';
export * from './interview-rating.interface';
export * from './job-application-score.interface';
export * from './response.interface';
