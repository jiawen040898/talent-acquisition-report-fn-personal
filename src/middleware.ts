import middy from '@middy/core';
import { initSentryMiddleware, loggerMiddleware } from '@pulsifi/fn';
import * as Sentry from '@sentry/serverless';
import { Handler } from 'aws-lambda';

import { version } from '../package.json';

initSentryMiddleware(version);

export const eventMiddleware = (handler: Handler) =>
    Sentry.AWSLambda.wrapHandler(middy(handler).use([loggerMiddleware()]));
