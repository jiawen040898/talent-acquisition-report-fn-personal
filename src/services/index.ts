export * from './chrome.service';
export * from './email.service';
export * from './feature-toggle.service';
export * from './file.service';
export * from './handlebars.service';
export * from './job-application-report.service';
export * from './psychology.service';
export * from './report-generator.service';
export * from './s3.service';
export * from './template.service';
