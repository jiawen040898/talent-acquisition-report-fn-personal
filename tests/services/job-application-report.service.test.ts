import { DataSource } from 'typeorm';

import * as database from '../../src/database';
import { JobApplicationReportService } from '../../src/services';
import { getTestDataSource, getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/job-application-report.service.test-data';

jest.mock('../../src/services/company.service', () => {
    const mBInstance = {
        getCompanyById: jest.fn(() => testData.companyResponse),
    };
    const mB = jest.fn(() => mBInstance);
    return { CompanyService: mB };
});

jest.mock('../../src/services/psychology.service', () => {
    const mBInstance = {
        getFrameworkDetails: jest.fn(() => testData.frameworkDetailResponse),
        getFitScoreRecipeById: jest.fn(
            () => testData.getFitScoreRecipeResponse,
        ),
    };
    const mB = jest.fn(() => mBInstance);
    return { PsychologyService: mB };
});

describe('JobApplicationReportService', () => {
    let dataSource: DataSource;
    const spygetDataSource = jest.spyOn(database, 'getDataSource');

    beforeAll(async () => {
        dataSource = await getTestDataSource();

        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesToBeAdded,
        );

        spygetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('prepareAssessmentReportSummary', () => {
        it('should get report summary for valid job application id', async () => {
            // Act
            const result =
                await JobApplicationReportService.prepareAssessmentReportSummary(
                    testData.mockExportCandidateReportRequest,
                    dataSource,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });
});
