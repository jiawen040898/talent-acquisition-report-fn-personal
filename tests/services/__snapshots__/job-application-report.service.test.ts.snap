// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobApplicationReportService prepareAssessmentReportSummary should get report summary for valid job application id 1`] = `
{
  "report_request": {
    "count": 1,
    "idempotence_key": "00000000-0000-0000-0000-000000000001",
    "index": 1,
    "job_application_id": "00000000-0000-0000-0000-000000000002",
    "variable": {
      "company_id": 5,
      "locale": "en-US",
      "notification_type": "email",
      "timezone": "Asia/Kuala_Lumpur",
      "user_account_id": 1,
      "user_email": "jaypete@gmail.com",
      "user_first_name": "Jay",
    },
  },
  "report_summary": {
    "address": "1 Pulsifi Road, Pulsifi Garden, 49990, KL.",
    "application_status": "applied",
    "assessment_descriptor_overview": {
      "thrive": [],
      "watch_out": [],
      "working_with": [],
    },
    "assessment_scores": {
      "culture_fit": [],
      "hard_skills": [
        {
          "match": true,
          "name": "Python",
          "score": 0.4610980749130249,
        },
        {
          "match": true,
          "name": "Angular",
          "score": 0.44767916202545166,
        },
        {
          "match": false,
          "name": "testing",
          "score": 0.26443392038345337,
        },
        {
          "match": true,
          "name": "NodeJs",
          "score": 0.22911888360977173,
        },
      ],
      "reasoning_average": undefined,
      "reasoning_logical": {
        "percentile": 0.77,
        "score": 0.85,
      },
      "reasoning_numeric": undefined,
      "reasoning_verbal": undefined,
      "role_fit": [
        {
          "ingredient_alias": "reasoning_logical",
          "ingredient_percentile": 0.77,
          "ingredient_score": null,
          "ingredient_weightage": 0.20826563051925115,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "reasoning_verbal",
          "ingredient_score": null,
          "ingredient_weightage": 0.19427764040974915,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "reasoning_average",
          "ingredient_score": null,
          "ingredient_weightage": 0.1907453196750265,
        },
        {
          "ingredient_alias": "reasoning_numeric",
          "ingredient_score": null,
          "ingredient_weightage": 0.16969268809607915,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "hard_skills",
          "ingredient_score": null,
          "ingredient_weightage": 0.16969268809607915,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "work_value",
          "ingredient_score": null,
          "ingredient_weightage": 0.14258801365830684,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "work_style",
          "ingredient_score": null,
          "ingredient_weightage": 0.14258801365830684,
          "ingredient_weighted_score": null,
        },
        {
          "ingredient_alias": "interest_riasec",
          "ingredient_score": null,
          "ingredient_weightage": 0.14258801365830684,
          "ingredient_weighted_score": null,
        },
      ],
      "work_experience": undefined,
      "work_interest": [],
      "work_style": [],
      "work_value": [],
    },
    "company_name": "Pulsifi",
    "contact_email": "jaypete@gmail.com",
    "contact_mobile_number": "",
    "culture_fit_framework_descriptor": "POLARIS",
    "culture_fit_framework_name": "POLARIS",
    "culture_fit_score": 8.2,
    "experiences": {
      "careers": [
        JobApplicationCareer {
          "end_date": 2023-10-01T00:00:00.000Z,
          "is_current": true,
          "organization": "Pulsifi",
          "place_formatted_address": "1 Pulsifi Road, Pulsifi Garden, 49990, KL.",
          "responsibilities_achievements": "This is my achievement and responsibility",
          "role": "QA",
          "start_date": 2023-01-01T00:00:00.000Z,
        },
        JobApplicationCareer {
          "end_date": 2022-10-01T00:00:00.000Z,
          "is_current": false,
          "organization": "Pulsifi",
          "place_formatted_address": "1 Pulsifi Road, Pulsifi Garden, 49990, KL.",
          "responsibilities_achievements": "This is my achievement and responsibility",
          "role": "QA",
          "start_date": 2022-01-01T00:00:00.000Z,
        },
      ],
      "educations": [
        JobApplicationEducation {
          "achievements": "This is my achievements",
          "degree_name": "degree_name",
          "end_date": 2023-10-01T00:00:00.000Z,
          "grade_cgpa": 0,
          "is_highest": true,
          "major_first": "major_first",
          "major_second": "major_second",
          "school_name": "school_name",
          "start_date": 2023-01-01T00:00:00.000Z,
        },
        JobApplicationEducation {
          "achievements": "This is my achievements",
          "degree_name": "degree_name",
          "end_date": 2022-10-01T00:00:00.000Z,
          "grade_cgpa": 0,
          "is_highest": false,
          "major_first": "major_first",
          "major_second": "major_second",
          "school_name": "school_name",
          "start_date": 2022-01-01T00:00:00.000Z,
        },
      ],
      "professional_summary": "This is my professional summary",
      "skills": [
        {
          "name": "Python",
          "proficiency": "expert",
          "source": "candidate",
        },
        {
          "name": "Angular",
          "proficiency": "expert",
          "source": "candidate",
        },
        {
          "name": "NodeJs",
          "proficiency": "expert",
          "source": "candidate",
        },
      ],
    },
    "files": [
      {
        "file_name": "Simon Peter Resume.pdf",
        "file_path": "https://pulsifi-sandbox-document.s3-ap-southeast-1.amazonaws.com/http://docs.pulsifi.me/cloud-url/resume.pdf",
      },
      {
        "file_name": "Simon Peter Attachment-0.pdf",
        "file_path": "https://pulsifi-sandbox-document.s3-ap-southeast-1.amazonaws.com/http://docs.pulsifi.me/cloud-url/attachment.pdf",
      },
    ],
    "first_name": "Simon",
    "has_assessment_completed": false,
    "has_passed_screening": true,
    "interview_rating_summary": {
      "company_name": "Pulsifi",
      "first_name": "Simon",
      "interview_average_rating": 4,
      "interview_response": [],
      "job_title": "Tester",
      "last_name": "Peter",
      "total_rating_submitted": 3,
    },
    "job_title": "Tester",
    "last_name": "Peter",
    "last_position": "QA at Pulsifi",
    "nationality": "Malaysian",
    "role_fit_score": 8.1,
    "screening_answers": [
      {
        "answer": "bachelor",
        "criteria_status": "pass",
        "question": "Education Level",
        "required": true,
        "tag": "met_grade",
      },
      {
        "answer": "bachelor",
        "criteria_status": "pass",
        "question": "Education Level",
        "required": true,
        "tag": "met_major",
      },
    ],
    "screening_tags": [
      "met_grade",
      "met_major",
    ],
    "source": "Pulsifi",
    "total_assessment": 2,
    "total_assessment_completed": 4,
  },
}
`;
